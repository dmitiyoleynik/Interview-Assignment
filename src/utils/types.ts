export type Ship = {
  id: number;
  built: number;
  name: string;
  lengthMeters: number;
  beamMeters: number;
  maxTEU: number;
  owner: string;
  grossTonnage: number | string;
};
