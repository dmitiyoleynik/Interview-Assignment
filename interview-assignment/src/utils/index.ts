import { useMediaQuery, useTheme } from '@mui/material';

export const getCountryOfOrigin = (shipOwner: string) => {
  const regex = /\(([^)]+)\)/;
  const matches = regex.exec(shipOwner);

  return matches && matches[1];
};

type size = 'xxs' | 'xs' | 'sm' | 'md' | 'lg';

type sizes<T> = {
  xxs: T;
  xs: T;
  sm: T;
  md: T;
  lg: T;
};

export const useFlexibleSizes = <T>(sizes: sizes<T>) => {
  const theme = useTheme();

  const breakpointsMap: { useMediaQueryResult: boolean; type: size }[] = [
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('lg')),
      type: 'lg',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('md')),
      type: 'md',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('sm')),
      type: 'sm',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up(280)),
      type: 'xs',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.down(280)),
      type: 'xxs',
    },
  ];
  console.log({ breakpointsMap });

  const matchingBreakpoint = breakpointsMap.find((s) => s.useMediaQueryResult);

  return matchingBreakpoint ? sizes[matchingBreakpoint.type] : sizes.xxs;
};
