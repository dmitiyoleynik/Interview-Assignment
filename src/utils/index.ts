import { useMediaQuery, useTheme } from '@mui/material';

export const getCountryOfOrigin = (shipOwner: string) => {
  const regex = /\(([^)]+)\)/;
  const matches = regex.exec(shipOwner);
  const matchIndex = 1;

  return matches?.[matchIndex];
};

enum Size {
  'xxs' = 'xxs',
  'xs' = 'xs',
  'sm' = 'sm',
  'md' = 'md',
  'lg' = 'lg',
}

type Sizes<T> = {
  [key in Size]: T;
};

export const useFlexibleSizes = <T>(sizes: Sizes<T>) => {
  const theme = useTheme();

  const breakpointsMap: { useMediaQueryResult: boolean; type: Size }[] = [
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('lg')),
      type: Size.lg,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('md')),
      type: Size.md,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('sm')),
      type: Size.sm,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up(280)),
      type: Size.xs,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.down(280)),
      type: Size.xxs,
    },
  ];
  console.log({ breakpointsMap });

  const matchingBreakpoint = breakpointsMap.find((s) => s.useMediaQueryResult);

  return matchingBreakpoint ? sizes[matchingBreakpoint.type] : sizes.xxs;
};
