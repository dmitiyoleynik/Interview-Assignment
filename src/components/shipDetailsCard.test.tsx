import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';

import ShipDetailsCard from 'src/components/shipDetailsCard';
import { Ship } from 'src/utils/types';

test('ShipDetailsCard should match snapshot', () => {
  const ship: Ship = {
    id: 3,
    built: 2017,
    name: 'OOCL Japan',
    lengthMeters: 399.87,
    beamMeters: 58.8,
    maxTEU: 21413,
    owner: 'OOCL (Hong Kong)',
    grossTonnage: 210890,
  };
  const component = renderer.create(
    <MemoryRouter>
      <ShipDetailsCard ship={ship} />
    </MemoryRouter>,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
