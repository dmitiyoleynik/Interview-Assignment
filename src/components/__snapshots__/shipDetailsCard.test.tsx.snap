// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShipDetailsCard should match snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-6ium34-MuiPaper-root"
  data-testid="ship-3"
>
  <h6
    className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
  >
    OOCL Japan
  </h6>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Country of origin: 
    Hong Kong
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Built: 
    2017
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Name: 
    OOCL Japan
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Length: 
    399.87
     meters
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Beam: 
    58.8
     meters
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Max TEU: 
    21413
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Owner: 
    OOCL (Hong Kong)
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Gross Tonnage: 
    210890
     tons
  </p>
  <a
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1njilb0-MuiButtonBase-root-MuiButton-root"
    href="/"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
  >
    Back to List
  </a>
</div>
`;
